File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/LogLineParserTest.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit;
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory;
import org.junit.Test;

import java.time.OffsetDateTime;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class LogLineParserTest {

    @Test
    public void parseCommitRaisesExceptionWhenAuthorIsMissing() {
        // given
        MetricsFactory metricsFactory = mock(MetricsFactory.class);
        LogParserStrategy parserStrategy = mock(LogParserStrategy.class);
        when(parserStrategy.parseAuthor(any())).thenReturn(Optional.empty());
        when(parserStrategy.parseDate(any())).thenReturn(Optional.of(OffsetDateTime.now()));
        when(parserStrategy.parseModifications(any())).thenReturn(Collections.emptyList());
        LogLineParser parser = new LogLineParser(parserStrategy, metricsFactory);

        // when & then
        assertThatThrownBy(() -> parser.parseCommit(Collections.emptyList())).isInstanceOf(RuntimeException.class);
    }

    @Test
    public void parseCommitRaisesExceptionWhenCommitDateIsMissing() {
        // given
        MetricsFactory metricsFactory = mock(MetricsFactory.class);
        LogParserStrategy parserStrategy = mock(LogParserStrategy.class);
        when(parserStrategy.parseAuthor(any())).thenReturn(Optional.of("An Author"));
        when(parserStrategy.parseDate(any())).thenReturn(Optional.empty());
        when(parserStrategy.parseModifications(any())).thenReturn(Collections.emptyList());
        LogLineParser parser = new LogLineParser(parserStrategy, metricsFactory);

        // when & then
        assertThatThrownBy(() -> parser.parseCommit(Collections.emptyList())).isInstanceOf(RuntimeException.class);
    }

    @Test
    public void parseCommit() {
        // given
        MetricsFactory metricsFactory = mock(MetricsFactory.class);
        LogParserStrategy parserStrategy = mock(LogParserStrategy.class);
        String author = "An Author";
        OffsetDateTime commitDate = OffsetDateTime.now();
        List<String> filenames = Arrays.asList("src/Main.java", "src/Util.java");
        List<String> input = Collections.emptyList();
        when(parserStrategy.parseAuthor(input)).thenReturn(Optional.of(author));
        when(parserStrategy.parseDate(input)).thenReturn(Optional.of(commitDate));
        when(parserStrategy.parseModifications(input))
                .thenReturn(filenames.stream().map(Modification::new).collect(Collectors.toList()));
        LogLineParser parser = new LogLineParser(parserStrategy, metricsFactory);

        // when
        Commit commit = parser.parseCommit(input);

        //then
        assertThat(commit.getAuthor()).isEqualTo(author);
        assertThat(commit.getFilenames()).isEqualTo(filenames);
        assertThat(commit.getCommitDate()).isEqualTo(commitDate);
    }
}

File: analysis/import/SCMLogParserV2/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparserv2/parser/LogLineParserTest.kt
package de.maibornwolff.codecharta.importer.scmlogparserv2.parser

import de.maibornwolff.codecharta.importer.scmlogparserv2.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.MetricsFactory
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.time.OffsetDateTime
import java.util.Arrays

class LogLineParserTest {
    private val metricsFactory = mockk<MetricsFactory>()

    @Before
    fun setup() {
        every { metricsFactory.createMetrics() } returns emptyList()
    }

    @Test
    fun parseCommit() {
        // given
        val parserStrategy = mockk<LogParserStrategy>()
        val author = "An Author"
        val commitDate = OffsetDateTime.now()
        val filenames = Arrays.asList("src/Main.java", "src/Util.java")
        val input = emptyList<String>()

        every { parserStrategy.parseAuthor(any()) } returns author
        every { parserStrategy.parseDate(any()) } returns commitDate
        every { parserStrategy.parseModifications(input) } returns filenames.map { Modification(it) }
        every { parserStrategy.parseIsMergeCommit(input) } returns false

        val parser = LogLineParser(parserStrategy, metricsFactory)

        // when
        val commit = parser.parseCommit(input)

        // then
        assertThat(commit.author).isEqualTo(author)
        assertThat(commit.filenames).isEqualTo(filenames)
        assertThat(commit.commitDate).isEqualTo(commitDate)
    }
}
